backend/src/email/services/email-transactional.service.ts

```diff
@@ -50,6 +54,53 @@ async function sendMessage({
   }
 }
 
+const rateLimitClients = {
+  user: new RateLimiterRedis({
+    storeClient: RedisService.rateLimitClient,
+    keyPrefix: 'transactionalEmail.user:',
+    points: config.get('transactionalEmail.userRate'),
+    duration: config.get('transactionalEmail.window'),
+  }),
+  global: new RateLimiterRedis({
+    storeClient: RedisService.rateLimitClient,
+    keyPrefix: 'transactionalEmail.global:',
+    points: config.get('transactionalEmail.globalRate'),
+    duration: config.get('transactionalEmail.window'),
+  }),
+}
+
+async function rateLimit(userKey: number, globalKey: string) {
+  const ACTION = 'rateLimit'
+
+  const userRateLimiterRes = await rateLimitClients.user.get(userKey)
+  const globalRateLimiterRes = await rateLimitClients.global.get(globalKey)
+  // If rateLimiterRes is null, the key has not been used before
+  const remainingUserTokens = userRateLimiterRes?.remainingPoints ?? 1
+  const remainingGlobalTokens = globalRateLimiterRes?.remainingPoints ?? 1
+  if (!remainingUserTokens || !remainingGlobalTokens) {
```

Comment:
I think this check does not guarantee that the `consume` will succeed later on. Assuming two requests that are processed in parallel, both of them will first see that they have remaining tokens. However, later on, only one `consume` will succeed. The get and set of the token value must be in one atomic operation. 

So this still does not solve the initial problem you had.

Reply:
After discussion, we decided that a global rate limit isn't that useful and is likely to have edge cases. Hence, I've change back to just using a per-user rate limit.